### PR TITLE
filters/diag: add histogram latency filters

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Add filters that add latency based on the configured histogram. See https://stats.stackexchange.com/questions/464006/generating-samples-from-a-histogram

Example:
```
$ bin/skipper -inline-routes='* -> histogramRequestLatency("5ms", 20, "10ms", 30, "15ms", 50, "20ms") -> status(204) -> <shunt>' -access-log-disabled
```

Latency distribution:
```
~$ echo "GET http://localhost:9090" | vegeta attack -rate=100/s -duration=20s | vegeta report -type hist[0ms,6ms,11ms,16ms,21ms,26ms]
Bucket         #    %       Histogram
[0s,    6ms]   8    0.40%
[6ms,   11ms]  372  18.60%  #############
[11ms,  16ms]  581  29.05%  #####################
[16ms,  21ms]  964  48.20%  ####################################
[21ms,  26ms]  75   3.75%   ##
[26ms,  +Inf]  0    0.00%
```
(note that buckets are shifted by 1ms to account for intrinsic latency)

### CODE REVIEW

File: filters/diag/histogram.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		if err != nil || max <= 0 {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		if i > 0 && max <= f.buckets[i/2-1] {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		f.buckets[i/2] = max
+	}
+
+	if f.sum > 0 {
```
Comment (MustafaSaber) [inline filters/diag/histogram.go:86]: shouldn't this return an error instead of going around it?

Comment (szuecs) [reply filters/diag/histogram.go:86]: We use relative weights in other cases (f.e. `Traffic()`), so should be fine

File: filters/diag/histogram.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		if upto >= r {
+			break
+		}
+	}
+	max := f.buckets[i]
+	var min time.Duration
+	if i > 0 {
+		min = f.buckets[i-1]
+	}
+	return min + time.Duration(rand.Int63n(int64(max-min))) // #nosec
```
Comment (szuecs) [inline filters/diag/histogram.go:122]: How does `min + x` match the example:
```
// Eskip example:
//
//	r: * -> histogramRequestLatency(50, "5ms", 0, "10ms", 30, "15ms", 20, "20ms") -> "https://www.example.org";
//
// The example above adds a latency
// * between implicit 0ms and 5ms to 50% of the requests   <--
```

Comment (AlexanderYastrebov) [reply filters/diag/histogram.go:122]: Previous loop selects `i`-th bucket according to its weight. This bucket left boundary `min` is `bucket[i-1]` (or zero) and right boundary `max` is `bucket[i]`. Then filter selects a random value from `[min, max)` interval as `rand(min, max) = min + rand(0, max-min)`

File: filters/diag/histogram.go (hunk 1 above)
Comment (AlexanderYastrebov) [reply filters/diag/histogram.go:86]: When sum is zero filter becomes a noop. Maybe we should disallow all zero weights and indeed return an error here.

Comment (AlexanderYastrebov) [reply filters/diag/histogram.go]: Fixed

File: filters/diag/histogram.go (hunk 2 above)
Comment (szuecs) [reply filters/diag/histogram.go:124]: Where is the implicit 0ms?

Comment (AlexanderYastrebov) [reply filters/diag/histogram.go:124]: The example is not wrong, please see docs for explanation:
```
The example above adds a latency
* between implicit 0ms and 5ms to 50% of the requests
* between 5ms and 10ms to 0% of the requests
* between 10ms and 15ms to 30% of the requests
* and between 15ms and 20ms to 20% of the requests.
```

Comment (AlexanderYastrebov) [reply filters/diag/histogram.go:124]: If it makes it easier to understand we may change filter to always require `"0ms"` (or any zero duration e.g. "0h") as a first argument.

Comment (AlexanderYastrebov) [reply filters/diag/histogram.go:124]: Or I we can actually swap order of weights and duration and require first and last arguments to be durations, e.g:
`histogramRequestLatency("5ms", 10, "10ms", 0, "20ms", 90, "30ms")` would add 5-10ms latency to 10% of requests and 20-30ms to 90%

Comment (szuecs) [reply filters/diag/histogram.go:124]: Order of args does not matter.

I think what needs to be clear is min. Either min is the upper bound of the leftmost bucket or the example or code is wrong.
Maybe min is not the right name for it. That's just what is not easy to understand and maybe you should define a min as absolute min?!

Comment (AlexanderYastrebov) [reply filters/diag/histogram.go:122]: Thank you.

> I think your last comment would make most sense as user input.

I've updated the code accordingly.

**Comment by szuecs** [comment]
:+1:

**Comment by MustafaSaber** [comment]
👍

**Pull request merged**